src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsUrlTest.java

```diff
@@ -0,0 +1,187 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2020 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.internal;
+
+import static com.google.common.truth.Truth.assertWithMessage;
+
+import java.io.File;
+import java.io.FilenameFilter;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import javax.xml.parsers.SAXParser;
+import javax.xml.parsers.SAXParserFactory;
+
+import org.junit.jupiter.api.Test;
+import org.xml.sax.Attributes;
+import org.xml.sax.helpers.DefaultHandler;
+
+public class XdocsUrlTest {
+
+    public static final String PACKAGE_NAME =
+            "src/main/java/com/puppycrawl/tools/checkstyle/checks";
+
+    public static final String SUFFIX_JAVA = ".java";
+
+    public static final String SUFFIX_CHECK_JAVA = "Check.java";
+
+    public static final String MISC = "misc";
+
+    public static final String ANNOTATION = "annotation";
+
+    public static final String COMMENTS_INDENTATION = "CommentsIndentation";
+
+    public static final String INDENTATION = "Indentation";
+
+    public static final String SUPPRESS_WARNINGS_HOLDER = "SuppressWarningsHolder";
+
+    public static final Path AVAILABLE_CHECKS_PATH = Paths.get("src/xdocs/checks.xml");
+
+    private static Map<String, String[]> getXdocsMap() {
+        final File checks = new File(PACKAGE_NAME);
+        final String[] checksSubDirNames = checks.list((rootDir, pathname) -> {
+            return !pathname.endsWith(SUFFIX_JAVA);
+        });
```

Comment:
I assume this is getting the list of checks in checkstyle and their packages. Why don't you just use `CheckUtil.getCheckstyleChecks` ?

Reply:
becase 1. `CheckUtil.getCheckstyleChecks` only return checks that are TreeWalkerCheck, but checks.xml contains all checks, such as HeaderCheck, LineLengthCheck，2. and `CheckUtil.getCheckstyleChecks` return a Set<Class<?>>, which is not classified according to packages so we also need to the similar work to classify it.

Reply:
1. There is `CheckUtil.getCheckstyleModules` which will return everything. You can use `ModuleReflectionUtil` to trim down to the specifics you need.
2. Since you get the class, you can obtain the package name from that and build using a Map and list.

By scanning files you are going to get more than final classes, you could pick up abstract classes like `AbstractSuperCheck` and any misc files we have to add.

Reply:
ok, done~